set/set_test.go

```diff
@@ -0,0 +1,365 @@
+	b = New[string]("2", "2", "1")
+	if !a.Equal(b) {
+		t.Errorf("Expected to be equal: %v vs %v", a, b)
+	}
+
+	// Edge cases around empty sets / empty strings
+	a = New[string]()
+	b = New[string]()
+	if !a.Equal(b) {
+		t.Errorf("Expected to be equal: %v vs %v", a, b)
+	}
+
+	b = New[string]("1", "2", "3")
+	if a.Equal(b) {
+		t.Errorf("Expected to be not-equal: %v vs %v", a, b)
+	}
+
+	b = New[string]("1", "2", "")
+	if a.Equal(b) {
+		t.Errorf("Expected to be not-equal: %v vs %v", a, b)
+	}
+
+	// Check for equality after mutation
+	a = New[string]()
+	a.Insert("1")
+	if a.Equal(b) {
+		t.Errorf("Expected to be not-equal: %v vs %v", a, b)
+	}
+
+	a.Insert("2")
+	if a.Equal(b) {
+		t.Errorf("Expected to be not-equal: %v vs %v", a, b)
+	}
+
+	a.Insert("")
+	if !a.Equal(b) {
+		t.Errorf("Expected to be equal: %v vs %v", a, b)
+	}
+
+	a.Delete("")
+	if a.Equal(b) {
+		t.Errorf("Expected to be not-equal: %v vs %v", a, b)
+	}
+}
+
+func TestStringUnion(t *testing.T) {
+	tests := []struct {
+		s1       Set[string]
+		s2       Set[string]
+		expected Set[string]
+	}{
+		{
+			New[string]("1", "2", "3", "4"),
+			New[string]("3", "4", "5", "6"),
+			New[string]("1", "2", "3", "4", "5", "6"),
+		},
+		{
+			New[string]("1", "2", "3", "4"),
+			New[string](),
+			New[string]("1", "2", "3", "4"),
+		},
+		{
+			New[string](),
+			New[string]("1", "2", "3", "4"),
+			New[string]("1", "2", "3", "4"),
+		},
+		{
+			New[string](),
+			New[string](),
+			New[string](),
+		},
+	}
+
+	for _, test := range tests {
+		union := test.s1.Union(test.s2)
+		if union.Len() != test.expected.Len() {
+			t.Errorf("Expected union.Len()=%d but got %d", test.expected.Len(), union.Len())
+		}
+
+		if !union.Equal(test.expected) {
+			t.Errorf("Expected union.Equal(expected) but not true.  union:%v expected:%v", union.SortedList(), test.expected.SortedList())
+		}
+	}
+}
+
+func TestStringIntersection(t *testing.T) {
+	tests := []struct {
+		s1       Set[string]
+		s2       Set[string]
+		expected Set[string]
+	}{
+		{
+			New[string]("1", "2", "3", "4"),
+			New[string]("3", "4", "5", "6"),
+			New[string]("3", "4"),
+		},
+		{
+			New[string]("1", "2", "3", "4"),
+			New[string]("1", "2", "3", "4"),
+			New[string]("1", "2", "3", "4"),
+		},
+		{
+			New[string]("1", "2", "3", "4"),
+			New[string](),
+			New[string](),
+		},
+		{
+			New[string](),
+			New[string]("1", "2", "3", "4"),
+			New[string](),
+		},
+		{
+			New[string](),
+			New[string](),
+			New[string](),
+		},
+	}
+
+	for _, test := range tests {
+		intersection := test.s1.Intersection(test.s2)
+		if intersection.Len() != test.expected.Len() {
+			t.Errorf("Expected intersection.Len()=%d but got %d", test.expected.Len(), intersection.Len())
+		}
+
+		if !intersection.Equal(test.expected) {
+			t.Errorf("Expected intersection.Equal(expected) but not true.  intersection:%v expected:%v", intersection.SortedList(), test.expected.SortedList())
+		}
+	}
+}
+
+func TestNewSetFromMapKeys(t *testing.T) {
```

Comment:
TestKeySet ?

Reply:
I changed this to be conformant with the apimachinery implementation